### DISCUSSION TITLE
Add extension to support letter lists.

### DISCUSSION THREAD

**Discussion by joapuiib** [facelessuser/pymdown-extensions#2454] [Ideas]
I think it would be nice to have an extension that supported all type of lists:

- Default (numbers).
- Upper case letters (`type="A"`)
    ```
    A. List item
    B. List item
    C. List item
    ```
- Lower case letters (`type="a"`)
    ```
    a. List item
    b. List item
    c. List item
    ```

- Upper case roman numbers (`type="I"`)
    ```
    I. List item
    II. List item
    III. List item
    ```
- Lower case roman numbers (`type="i"`)
    ```
    i. List item
    ii. List item
    iii. List item
    ```

Also, could be nice to be able to set the `start` number/letter, just as [`sane_lists`](https://python-markdown.github.io/extensions/sane_lists/) does.

```
C. List item
D. List item

m. List item
n. List item

III. List item
IV. List item

ix. List item
x. List item
```

**Comment by facelessuser** [comment]
See issue #2418

**Comment by joapuiib** [comment]
Since it is something I'm planning to use, I have written a small extension with the previous use cases.

Just in case it is useful to someone, I'll leave it here: https://github.com/joapuiib/pymdown-fancylists

Thanks!

**Reply by facelessuser** [reply]
There are likely some edge cases you may or may not have considered. When do you accept `v.` as a roman numeral list and when do you accept `v.` as an alpha list? What do you do if you get something that looks like roman numerals but is nonsense: `iviv`.

This should be an alpha list.

```
a. item 1
i. item2
```

This should be a roman numeral list:

```
i. item 2
ii. item 2
```

This is ambiguous, but references seem to assume alpha in this case:

```
v. item2
```

I'm exploring the possibility via prototype, but I am not sure if I'm all in yet.

One thing to note, this behaves very Pandoc-like currently in that:

- It supports lists of type `#.`, `1.`, `a.`, `i.`, `#)`, `1)`, `a)`,  `i)`, `(#)`, `(1)`, `(a)`,  and `(i)`. Also capitalized versions.
- Each list type is treated as its own list type, both uppercase and lowercase are treated unique. If starting a new list type, a new list is created and not appended to the previous list.
- The number (or symbol) indicating the start of the list is taken into account and the list will start at that value.
- `pymdownx.saneheaders` is required to be enabled for `#` style lists to work as otherwise they will be identified as headers.
- I'm not sure if I'm 100% sold on `()` style lists or `#` style lists, but implemented everything to understand how it would behave.
- Alpha lists are currently contained to `a` and do not allow `aa`.
- Obviously, there are rules to determine when some roman numerals are assumed roman numerals or alpha lists. Generally new lists are determined to be roman if they start with `i` or `I`  or are wider than one character.
- Roman numerals are checked for validity and are ignored if they are not valid.
- NONE OF THIS IS STRICTLY TESTED CURRENTLY. I've made assertions above, but I have not extensively tested any of this.
- Lastly, all of this is subject to change if/when I decide to support this.

Here is the prototype: https://gist.github.com/facelessuser/a6613237425b78e843c32268ef464e2e

**Reply by joapuiib** [reply]
I had considered those cases, but since I'm not familiar with writing extensions, I started with baby steps and work with what I need right now.

Obviously, your prototype is way more mature and provides more features than what I've done.

I'm willing to help you and provide tests for your code, either you want to include it in [pymdown-extensions](https://github.com/facelessuser/pymdown-extensions) or in a separate package. I think there's no reason to have two different implementations/packages.

**Reply by facelessuser** [reply]
If I decide to support some iteration of this, it will be in pymdown-extensions, not some separate package. I'm simply in the prototype stage. I want to see what limitations there are, what potential issues could manifest, how fragile it is. All of this will feed into whether I choose to move forward with it. I'm also evaluating the various parts of the referenced "fancylist" in Pandoc.  Do we monkey everything? Or do pick specific parts? Do we extend anything?

Anyway, I'm actively considering it, I just don't know what my decision is yet.

**Reply by facelessuser** [reply]
I think I will not be keeping `(symbol)` method if I move forward. I only had implemented `(digit)` in the prototype, and things like GitHub don't even support such syntax, they support `symbol.` and `symbol)`. I don't want to complicate patterns to cover this case and my intention is not to provide a 1:1 implementation of Pandoc either.

Pandoc also adds a requirement for single letter uppercase list markers to have two spaces after the marker. I will likely add this requirement as well. I'm sure there are reasons.

**Reply by joapuiib** [reply]
> Pandoc also adds a requirement for single letter uppercase list markers to have two spaces after the marker. I will likely add this requirement as well. I'm sure there are reasons.

As [Pandoc documentation](https://pandoc.org/chunkedhtml-demo/8.7-lists.html#fn1) says:

> The point of this rule is to ensure that normal paragraphs starting with people’s initials, like
> 
> B. Russell was an English philosopher.
> do not get treated as list items.
> 
> This rule will not prevent
> 
> (C) 2007 Joe Smith
> from being interpreted as a list item. In this case, a backslash escape can be used:
> 
> (C\\) 2007 Joe Smith

**Reply by facelessuser** [reply]
This is what I'm thinking so far. By default, FancyList would extend list formats such that `1.` and `1)` would work. These formats would be added with functionality similar to [Sane Lists](https://python-markdown.github.io/extensions/sane_lists/) but with the added behavior that swapping list types `1.` vs `1)` would force a new list.

All other formats are probably not supported syntactically in many editors. It's possible not everyone wants such behavior, and might actually control this automatically with CSS when lists are nested, but that's okay. I think additional list formats would be opt-in. You could enable one or all of them if you desire.

Someone using MkDocs might do this to enable all formats; decimal cannot be toggled and is always enabled.

```yaml
markdown_extensions:
  - pymdownx.fancylists:
      addtional_ordered_styles:
      - alpha
      - roman
      - generic
 ```

**Reply by facelessuser** [reply]
Ugh, I meant to post this in the issue. If we can move conversation there it would be appreciated.
